### ISSUE TITLE
deadlock with read-transaction and write transaction on different go-routines

### ISSUE THREAD

**Issue by flywukong** [etcd-io/bbolt#854]
I don't quite understand why deadlock conflicts occur between read transactions and write transactions.

Goroutine 1: call View API, Read transaction, acquires metaLock during txn begin process, then fails to acquire mmap read lock
Goroutine 2:  call View API,  it is Read transaction, fails to acquire metaLock
Goroutine 3:   call Update API.  it is  Write transaction, fails to acquire mmap write lock during txCommit process

below is the stack 

```

goroutine 8961 [sync.Mutex.Lock]:
sync.runtime_SemacquireMutex(0xc01069bdf0?, 0xc0?, 0xc014bc0fa8?)
        runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0004c49b8)
        sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
        sync/mutex.go:90
github.com/etcd-io/bbolt.(*DB).beginTx(0xc0004c4800)
        github.com/etcd-io/bbolt@v1.3.3/db.go:551 +0x46 
github.com/etcd-io/bbolt.(*DB).Begin(0xc010ad2a50?, 0x0?)
        github.com/etcd-io/bbolt@v1.3.3/db.go:544 +0x25
github.com/etcd-io/bbolt.(*DB).View(0x30?, 0xc014bc1108)
        github.com/etcd-io/bbolt@v1.3.3/db.go:709 +0x30
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*Database).Has(0x65ec00?, {0xc01195d080?, 0xc014bc1190?, 0x29?})
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:245 +0x5d
github.com/ethereum/go-ethereum/core/rawdb.HasHeader({0x7f7bb8630308, 0xc001280ae0}, {0x46, 0xb, 0x2d, 0x2f, 0xe8, 0xad, 0xd0, 0xed, ...}, ...)
        github.com/ethereum/go-ethereum/core/rawdb/accessors_chain.go:368 +0x1f3
github.com/ethereum/go-ethereum/core.(*HeaderChain).HasHeader(0xc010de06c0, {0x46, 0xb, 0x2d, 0x2f, 0xe8, 0xad, 0xd0, 0xed, 0xa1, ...}, ...)
        github.com/ethereum/go-ethereum/core/headerchain.go:535 +0xc5
github.com/ethereum/go-ethereum/core.(*BlockChain).HasHeader(...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:86


        
goroutine 11414 [sync.RWMutex.Lock]:
sync.runtime_SemacquireRWMutex(0x2278140?, 0xd2?, 0x442d2b?)
        runtime/sema.go:87 +0x25
sync.(*RWMutex).Lock(0x65da6519c8354935?)
        sync/rwmutex.go:151 +0x6a
github.com/etcd-io/bbolt.(*DB).mmap(0xc0004c4800, 0x100000)
        github.com/etcd-io/bbolt@v1.3.3/db.go:329 +0x55
github.com/etcd-io/bbolt.(*DB).allocate(0xc0004c4800, 0xd6, 0x5)
        github.com/etcd-io/bbolt@v1.3.3/db.go:939 +0x10e
github.com/etcd-io/bbolt.(*Tx).allocate(0xc014863420, 0x5)
        github.com/etcd-io/bbolt@v1.3.3/tx.go:498 +0x36
github.com/etcd-io/bbolt.(*node).spill(0xc0003ec4d0)
        github.com/etcd-io/bbolt@v1.3.3/node.go:368 +0x289
github.com/etcd-io/bbolt.(*node).spill(0xc0003ec3f0)
        github.com/etcd-io/bbolt@v1.3.3/node.go:350 +0x9e
github.com/etcd-io/bbolt.(*Bucket).spill(0xc0148d6240)
        github.com/etcd-io/bbolt@v1.3.3/bucket.go:568 +0x365
github.com/etcd-io/bbolt.(*Bucket).spill(0xc014863438)
        github.com/etcd-io/bbolt@v1.3.3/bucket.go:535 +0x105
github.com/etcd-io/bbolt.(*Tx).Commit(0xc014863420)
        github.com/etcd-io/bbolt@v1.3.3/tx.go:160 +0xe5
github.com/etcd-io/bbolt.(*DB).Update(0xc0117b01e0?, 0xc010804fc0)
        github.com/etcd-io/bbolt@v1.3.3/db.go:701 +0xd2
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*batch).Write.func2()
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:565 +0x45
created by github.com/ethereum/go-ethereum/ethdb/bboltdb.(*batch).Write in goroutine 11480
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:564 +0x170


goroutine 9391 [sync.RWMutex.RLock]:
sync.runtime_SemacquireRWMutexR(0xc0128c7058?, 0x78?, 0xc0128c7088?)
        runtime/sema.go:82 +0x25
sync.(*RWMutex).RLock(...)
        sync/rwmutex.go:70
github.com/etcd-io/bbolt.(*DB).beginTx(0xc0004c4800)
        github.com/etcd-io/bbolt@v1.3.3/db.go:556 +0x74
github.com/etcd-io/bbolt.(*DB).Begin(0xc010ad2a50?, 0x0?)
        github.com/etcd-io/bbolt@v1.3.3/db.go:544 +0x25
github.com/etcd-io/bbolt.(*DB).View(0x30?, 0xc0128c7108)
        github.com/etcd-io/bbolt@v1.3.3/db.go:709 +0x30
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*Database).Has(0x65ec00?, {0xc014658630?, 0xc0128c7190?, 0x29?})
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:245 +0x5d
github.com/ethereum/go-ethereum/core/rawdb.HasHeader({0x7f7bb8630308, 0xc001280ae0}, {0x46, 0xb, 0x2d, 0x2f, 0xe8, 0xad, 0xd0, 0xed, ...}, ...)
        github.com/ethereum/go-ethereum/core/rawdb/accessors_chain.go:368 +0x1f3
github.com/ethereum/go-ethereum/core.(*HeaderChain).HasHeader(0xc010de06c0, {0x46, 0xb, 0x2d, 0x2f, 0xe8, 0xad, 0xd0, 0xed, 0xa1, ...}, ...)
        github.com/ethereum/go-ethereum/core/headerchain.go:535 +0xc5
github.com/ethereum/go-ethereum/core.(*BlockChain).HasHeader(...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:86
github.com/ethereum/go-ethereum/core.(*BlockChain).HasBlock(0xc010e08008, {0x46, 0xb, 0x2d, 0x2f, 0xe8, 0xad, 0xd0, 0xed, 0xa1, ...}, ...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:167 +0x6f
github.com/ethereum/go-ethereum/eth.(*ethHandler).handleBlockAnnounces(0xc0106cf4a0, 0xc010d3f3b0, {0xc014755440, 0x1, 0xc0128c74f8?}, {0xc014dce6b0, 0x1, 0xfbfe61b81876b9a1?})
        github.com/ethereum/go-ethereum/eth/handler_eth.go:104 +0x216
github.com/ethereum/go-ethereum/eth.(*ethHandler).Handle(0xc0106cf4a0, 0xc010d3f3b0, {0x3534be8?, 0xc011a31530})
        github.com/ethereum/go-ethereum/eth/handler_eth.go:65 +0x375
github.com/ethereum/go-ethereum/eth/protocols/eth.handleNewBlockhashes({0x3551488, 0xc0106cf4a0}, {0x3537280, 0xc000a814a0}, 0xc010d3f3b0)
        github.com/ethereum/go-ethereum/eth/protocols/eth/handlers.go:301 +0x21a
github.com/ethereum/go-ethereum/eth/protocols/eth.handleMessage({0x3551488, 0xc0106cf4a0}, 0xc010d3f3b0)
        github.com/ethereum/go-ethereum/eth/protocols/eth/handler.go:208 +0x570
```

**Comment by ahrtr** [comment]
I see that you are using bbolt 1.3.3, can you try the latest version 1.3.11 or main branch?

**Comment by flywukong** [comment]
> I see that you are using bbolt 1.3.3, can you try the latest version 1.3.11 or main branch?

updated ,  the result is same.  I also find  an iterator which contain a read txn and not release for long time in another go rountine  , it is may be similar like this issues https://github.com/boltdb/bolt/issues/378 ?
```

goroutine 10336 [sync.Mutex.Lock]:
sync.runtime_SemacquireMutex(0xc0006baf50?, 0xa0?, 0xc00146efa8?)
        runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0007a81c8)
        sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
        sync/mutex.go:90
go.etcd.io/bbolt.(*DB).beginTx(0xc0007a8008) 
        go.etcd.io/bbolt@v1.3.11/db.go:730 +0x46
go.etcd.io/bbolt.(*DB).Begin(0xc011736b10?, 0x0?)
        go.etcd.io/bbolt@v1.3.11/db.go:723 +0x25
go.etcd.io/bbolt.(*DB).View(0x30?, 0xc00146f108)
        go.etcd.io/bbolt@v1.3.11/db.go:901 +0x30
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*Database).Has(0x65ec00?, {0xc010bfc2a0?, 0xc00146f190?, 0x29?})
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:259 +0x5d
github.com/ethereum/go-ethereum/core/rawdb.HasHeader({0x7f72787bc3b8, 0xc000203c50}, {0x10, 0xa5, 0xa, 0x12, 0xdb, 0x75, 0xcf, 0x6, ...}, ...)
        github.com/ethereum/go-ethereum/core/rawdb/accessors_chain.go:368 +0x1f3
github.com/ethereum/go-ethereum/core.(*HeaderChain).HasHeader(0xc001234fc0, {0x10, 0xa5, 0xa, 0x12, 0xdb, 0x75, 0xcf, 0x6, 0x5e, ...}, ...)
        github.com/ethereum/go-ethereum/core/headerchain.go:535 +0xc5
github.com/ethereum/go-ethereum/core.(*BlockChain).HasHeader(...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:86
github.com/ethereum/go-ethereum/core.(*BlockChain).HasBlock(0xc01084a008, {0x10, 0xa5, 0xa, 0x12, 0xdb, 0x75, 0xcf, 0x6, 0x5e, ...}, ...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:167 +0x6f
github.com/ethereum/go-ethereum/eth.(*ethHandler).handleBlockAnnounces(0xc0003ddce0, 0xc0013ecc30, {0xc011788ea0, 0x1, 0xc00146f4f8?}, {0xc013381568, 0x1, 0x115a3abc9c677e5e?})
        github.com/ethereum/go-ethereum/eth/handler_eth.go:104 +0x216

goroutine 9545 [sync.RWMutex.Lock]:
sync.runtime_SemacquireRWMutex(0x446e78?, 0xa8?, 0xbf00c00100ebd0?)
        runtime/sema.go:87 +0x25
sync.(*RWMutex).Lock(0xc01054d6f8?)
        sync/rwmutex.go:151 +0x6a
go.etcd.io/bbolt.(*DB).mmap(0xc0007a8008, 0x101000)
        go.etcd.io/bbolt@v1.3.11/db.go:433 +0x4f
go.etcd.io/bbolt.(*DB).allocate(0xc0007a8008, 0xcb, 0x5)
        go.etcd.io/bbolt@v1.3.11/db.go:1132 +0x10e
go.etcd.io/bbolt.(*Tx).allocate(0xc0014516c0, 0x5)
        go.etcd.io/bbolt@v1.3.11/tx.go:417 +0x36
go.etcd.io/bbolt.(*node).spill(0xc0006f69a0)
        go.etcd.io/bbolt@v1.3.11/node.go:368 +0x212
go.etcd.io/bbolt.(*node).spill(0xc0006f68c0)
        go.etcd.io/bbolt@v1.3.11/node.go:350 +0x9e
go.etcd.io/bbolt.(*Bucket).spill(0xc013920d00)
        go.etcd.io/bbolt@v1.3.11/bucket.go:592 +0x365
go.etcd.io/bbolt.(*Bucket).spill(0xc0014516d8)
        go.etcd.io/bbolt@v1.3.11/bucket.go:559 +0x105
go.etcd.io/bbolt.(*Tx).Commit(0xc0014516c0)
        go.etcd.io/bbolt@v1.3.11/tx.go:164 +0xf9
go.etcd.io/bbolt.(*DB).Update(0xc010d377d0?, 0xc0105c6fc0)
        go.etcd.io/bbolt@v1.3.11/db.go:893 +0xd2
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*batch).Write.func2()
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:580 +0x45
created by github.com/ethereum/go-ethereum/ethdb/bboltdb.(*batch).Write in goroute

goroutine 9608 [sync.Mutex.Lock]:
sync.runtime_SemacquireMutex(0xc0006baf50?, 0xa0?, 0xc010c8efa8?)
        runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0007a81c8)
        sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
        sync/mutex.go:90
go.etcd.io/bbolt.(*DB).beginTx(0xc0007a8008)
        go.etcd.io/bbolt@v1.3.11/db.go:730 +0x46
go.etcd.io/bbolt.(*DB).Begin(0xc011736b10?, 0x0?)
        go.etcd.io/bbolt@v1.3.11/db.go:723 +0x25
go.etcd.io/bbolt.(*DB).View(0x30?, 0xc010c8f108)
        go.etcd.io/bbolt@v1.3.11/db.go:901 +0x30
github.com/ethereum/go-ethereum/ethdb/bboltdb.(*Database).Has(0x65ec00?, {0xc01093e360?, 0xc010c8f190?, 0x29?})
        github.com/ethereum/go-ethereum/ethdb/bboltdb/bblot.go:259 +0x5d
github.com/ethereum/go-ethereum/core/rawdb.HasHeader({0x7f72787bc3b8, 0xc000203c50}, {0x84, 0xd2, 0xf8, 0xb5, 0x50, 0xaf, 0xff, 0x47, ...}, ...)
        github.com/ethereum/go-ethereum/core/rawdb/accessors_chain.go:368 +0x1f3
github.com/ethereum/go-ethereum/core.(*HeaderChain).HasHeader(0xc001234fc0, {0x84, 0xd2, 0xf8, 0xb5, 0x50, 0xaf, 0xff, 0x47, 0xbe, ...}, ...)
        github.com/ethereum/go-ethereum/core/headerchain.go:535 +0xc5
github.com/ethereum/go-ethereum/core.(*BlockChain).HasHeader(...)
        github.com/ethereum/go-ethereum/core/blockchain_reader.go:86
```

**Comment by ahrtr** [comment]
Can you dump the call stack for all your goroutines?

Do you have any long running readonly transaction?

**Comment by flywukong** [comment]
> long running readonly transaction

yes,  we have two  long running readonly transactions, I am planning to avoid this solution

**Comment by ahrtr** [comment]
> we have two long running readonly transactions

A long running readonly transaction will block writing transaction when it needs to increase the size of the db file.



> Can you dump the call stack for all your goroutines?

But it is not clear why there are two goroutines being blocked on `db.metalock.Lock() `.  It's exactly the reason why we need to take a look at the call stacks for all your goroutines.

https://github.com/etcd-io/bbolt/blob/d128a10000a9d394686cf45be262a4fe966b03c4/db.go#L730

**Comment by ahrtr** [comment]
Closing since there is no feedback. Also there was long running readonly transaction, which would block rw transaction.

Please raise a new ticket if you still see this issue after removing the long running readonly transaction.
